### PR TITLE
Remove some code duplication between SequentialWriter and SequentialCompressionWriter

### PR DESCRIPTION
**Pull request by jaisontj**
[Edited] Major changes:
- `SequentialCompressionWriter` now inherits from `SequentialWriter`
- New method `get_writeable_message` inside of `SequentialWriter` to be called to get the message to be written to storage or cache

### CODE REVIEW

**Review by emersonknapp — APPROVED** [review]
This LGTM - but it looks like it will need another pass to consolidate the logic in `write`, so that the caching logic is not duplicated. In this case, it seems like the SequentialCompressionWriter isn't doing caching at all, that's how badly it's diverged.

File: rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp [hunk 1]
```diff
@@ -103,46 +99,7 @@ void SequentialCompressionWriter::open(
...
-    max_bagfile_size_ < storage_->get_minimum_split_file_size())
-  {
-    std::stringstream error;
-    error << "Invalid bag splitting size given. Please provide a value greater than " <<
-      storage_->get_minimum_split_file_size() << ". Specified value of " <<
-      storage_options.max_bagfile_size;
-    throw std::runtime_error{error.str()};
-  }
-
+  prepare_to_open(storage_options, converter_options);
```
Comment (emersonknapp) [inline rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:102]: can this method be deleted by overriding `prepare_to_open`, calling the superclass, then calling `setup_compression`?

Comment (jaisontj) [reply rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:102]: I don't particularly see the difference except in that `setup_compression` is maybe part of the `prepare` stage of this? I don't have a problem implementing this but am interested to know why you think this is better.

Comment (dabonnie) [reply rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:102]: +1 to @emersonknapp's comment. This leaves the superclass alone and implements the compression specific details in the relevant place.

**Review by dabonnie — CHANGES_REQUESTED** [review]
Can you please add unit tests for new methods?

**Comment by jaisontj** [comment]
> Can you please add unit tests for new methods?

The new methods are non public and used internally as helper methods. Do we usually write unit tests for those?

**Comment by emersonknapp** [comment]
No, generally you unit test the public interface, not the implementation details (private methods). You can of course unit test protected methods by making a test-specific subclass of the given class, but that depends on how much you consider it an interface surface vs an implementation detail.

**Review by dabonnie — APPROVED** [review]
Thanks for iterating, this is much cleaner!

**Comment by jaisontj** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12594/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7560/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10305/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12518/)

**Pull request merged**